x/staking/keeper/delegation.go

```diff
@@ -475,156 +528,147 @@ func (k Keeper) unbond(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValA
-		minTime = validator.UnbondingMinTime
+		completionTime = validator.UnbondingMinTime
 		height = validator.UnbondingHeight
-		return minTime, height, false
+		return completionTime, height, false
 
 	default:
 		panic("unknown validator status")
 	}
 }
 
 // begin unbonding an unbonding record
-func (k Keeper) BeginUnbonding(ctx sdk.Context,
-	delAddr sdk.AccAddress, valAddr sdk.ValAddress, sharesAmount sdk.Dec) (types.UnbondingDelegation, sdk.Error) {
-
-	// TODO quick fix, instead we should use an index, see https://github.com/cosmos/cosmos-sdk/issues/1402
-	_, found := k.GetUnbondingDelegation(ctx, delAddr, valAddr)
-	if found {
-		return types.UnbondingDelegation{}, types.ErrExistingUnbondingDelegation(k.Codespace())
-	}
+func (k Keeper) BeginUnbonding(ctx sdk.Context, delAddr sdk.AccAddress,
+	valAddr sdk.ValAddress, sharesAmount sdk.Dec) (completionTime time.Time, sdkErr sdk.Error) {
 
 	// create the unbonding delegation
-	minTime, height, completeNow := k.getBeginInfo(ctx, valAddr)
+	completionTime, height, completeNow := k.getBeginInfo(ctx, valAddr)
 
 	returnAmount, err := k.unbond(ctx, delAddr, valAddr, sharesAmount)
 	if err != nil {
-		return types.UnbondingDelegation{}, err
+		return completionTime, err
 	}
 	balance := sdk.NewCoin(k.BondDenom(ctx), returnAmount)
 
 	// no need to create the ubd object just complete now
 	if completeNow {
 		_, _, err := k.bankKeeper.AddCoins(ctx, delAddr, sdk.Coins{balance})
 		if err != nil {
-			return types.UnbondingDelegation{}, err
+			return completionTime, err
 		}
-		return types.UnbondingDelegation{MinTime: minTime}, nil
+		return completionTime, nil
 	}
 
-	ubd := types.UnbondingDelegation{
-		DelegatorAddr:  delAddr,
-		ValidatorAddr:  valAddr,
-		CreationHeight: height,
-		MinTime:        minTime,
-		Balance:        balance,
-		InitialBalance: balance,
-	}
-	k.SetUnbondingDelegation(ctx, ubd)
-	k.InsertUnbondingQueue(ctx, ubd)
+	ubd := k.SetUnbondingDelegationEntry(ctx, delAddr,
+		valAddr, height, completionTime, balance)
 
-	return ubd, nil
+	k.InsertUBDQueue(ctx, ubd, completionTime)
+	return completionTime, nil
 }
 
-// complete unbonding an unbonding record
-// CONTRACT: Expects unbonding passed in has finished the unbonding period
-func (k Keeper) CompleteUnbonding(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) sdk.Error {
+// CompleteUnbonding completes the unbonding of all mature entries in the
+// retrieved unbonding delegation object.
+func (k Keeper) CompleteUnbonding(ctx sdk.Context, delAddr sdk.AccAddress,
+	valAddr sdk.ValAddress) sdk.Error {
 
 	ubd, found := k.GetUnbondingDelegation(ctx, delAddr, valAddr)
 	if !found {
 		return types.ErrNoUnbondingDelegation(k.Codespace())
 	}
 
-	_, _, err := k.bankKeeper.AddCoins(ctx, ubd.DelegatorAddr, sdk.Coins{ubd.Balance})
-	if err != nil {
-		return err
+	ctxTime := ctx.BlockHeader().Time
+
+	// loop through all the entries and complete unbonding mature entries
+	for i := 0; i < len(ubd.Entries); i++ {
```

Comment:
I actually originally implemented with range, but ran into issue because we're removing entries, might be possible to use `range` still, but I think this is less confusing in this case